File: phoenix/test/marketplace/product_controller_test.exs
defmodule CardsProjectWeb.Marketplace.ProductControllerTest do
  use ExUnit.Case, async: false
  use CardsProjectWeb.ConnCase

  @valid_params %{
    "name" => "test",
    "price" => 1,
    "stock" => 0,
    "active" => true,
    "discount_percent" => 0,
    "featured" => true,
    "product_type" => "SingleCard"
  }

  describe "GET /api/products" do
    test "returns 200 with list", %{conn: conn} do
      conn = get(conn, "/api/products")
      assert json_response(conn, 200) |> is_list()
    end
  end

  describe "GET /api/products?q=test" do
    test "returns 200 with search results", %{conn: conn} do
      conn = get(conn, "/api/products?q=test")
      assert json_response(conn, 200) |> is_list()
    end
  end

  describe "POST /api/products" do
    test "creates record and returns 201", %{conn: conn} do
      conn = post(conn, "/api/products", @valid_params)
      assert %{"id" => _id} = json_response(conn, 201)
    end
    test "fails when price_positive violated", %{conn: conn} do
      params = Map.put(@valid_params, "price", 0)
      conn = post(conn, "/api/products", params)
      assert conn.status in [400, 422]
    end
    test "fails when stock_not_negative violated", %{conn: conn} do
      params = Map.put(@valid_params, "stock", -1)
      conn = post(conn, "/api/products", params)
      assert conn.status in [400, 422]
    end
    test "fails when discount_percent_range violated", %{conn: conn} do
      params = Map.put(@valid_params, "discount_percent", 1001)
      conn = post(conn, "/api/products", params)
      assert conn.status in [400, 422]
    end
  end

  describe "GET /api/products/:id" do
    test "returns 200 for existing record", %{conn: conn} do
      {:ok, record} = CardsProject.Marketplace.create_product(@valid_params)
      conn = get(conn, "/api/products/#{record.id}")
      assert json_response(conn, 200)
    end
  end

  @patch_params %{
    "featured" => true
  }

  describe "PUT /api/products/:id" do
    test "updates and returns 200", %{conn: conn} do
      {:ok, record} = CardsProject.Marketplace.create_product(@valid_params)
      conn = put(conn, "/api/products/#{record.id}", @patch_params)
      assert json_response(conn, 200)
    end
  end

end

File: spring/src/test/java/cardsproject/controller/marketplace/ProductControllerTest.java
package cardsproject.controller.marketplace;

import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.http.MediaType;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.transaction.annotation.Transactional;

import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.*;

@SpringBootTest
@AutoConfigureMockMvc
@Transactional
public class ProductControllerTest {

    @Autowired
    private MockMvc mockMvc;

    @Test
    void list_returns200() throws Exception {
        mockMvc.perform(get("/api/products"))
            .andExpect(status().isOk());
    }

    @Test
    void create_returns201() throws Exception {
        mockMvc.perform(post("/api/products")
            .contentType(MediaType.APPLICATION_JSON)
            .content("{ \"name\": \"test\", \"price\": 0.01 }"))
            .andExpect(status().isCreated());
    }

    @Test
    void show_returns200or404() throws Exception {
        mockMvc.perform(get("/api/products/1"))
            .andExpect(result -> {
                int status = result.getResponse().getStatus();
                assert status == 200 || status == 404;
            });
    }

    @Test
    void delete_returns204or404() throws Exception {
        mockMvc.perform(delete("/api/products/1"))
            .andExpect(result -> {
                int status = result.getResponse().getStatus();
                assert status == 204 || status == 404 || status == 500 || status == 501;
            });
    }
    @Test
    void create_fails_when_price_positive_violated() throws Exception {
        // Product price must be greater than zero → 400 (Bean Validation)
        mockMvc.perform(post("/api/products")
            .contentType(MediaType.APPLICATION_JSON)
            .content("{ \"name\": \"test\", \"productType\": \"SINGLECARD\", \"stock\": 1, \"active\": true, \"discountPercent\": 1, \"featured\": true, \"price\": 0.00 }"))
            .andExpect(status().isBadRequest());
    }

    @Test
    void create_fails_when_stock_not_negative_violated() throws Exception {
        // Product stock must not be negative → 400 (Bean Validation)
        mockMvc.perform(post("/api/products")
            .contentType(MediaType.APPLICATION_JSON)
            .content("{ \"name\": \"test\", \"productType\": \"SINGLECARD\", \"price\": 0.00, \"active\": true, \"discountPercent\": 1, \"featured\": true, \"stock\": -1 }"))
            .andExpect(status().isBadRequest());
    }

    @Test
    void create_fails_when_discount_percent_range_violated() throws Exception {
        // Product discount percent must be between 0 and 100 → 400 (Bean Validation)
        mockMvc.perform(post("/api/products")
            .contentType(MediaType.APPLICATION_JSON)
            .content("{ \"name\": \"test\", \"productType\": \"SINGLECARD\", \"price\": 0.00, \"stock\": 1, \"active\": true, \"featured\": true, \"discountPercent\": 101 }"))
            .andExpect(status().isBadRequest());
    }
}
